### ISSUE TITLE
(Unstable) API panic since last update

### ISSUE THREAD

**Issue by Break-Off** [go-vikunja/api#72]
### Description

I have a self-hosted Vikunja instance that updated overnight and since then it won't start:

```
info: creating the new user vikunja with 1000:1000
usermod: no changes
2023/05/25 17:44:29 No config file found, using default or config from environment variables.
2023-05-25T17:44:29.109566577Z: CRITICAL        ▶ migration/Migrate 002 Migration failed: dial tcp 172.23.0.3:3306: connect: connection refused
info: creating the new user vikunja with 1000:1000
usermod: no changes
2023/05/25 17:44:30 No config file found, using default or config from environment variables.
2023-05-25T17:44:31.253807025Z: WARNING ▶ [DATABASE] 054 Table projects has column identifier but struct has not related field
2023-05-25T17:44:31.253902728Z: WARNING ▶ [DATABASE] 055 Table projects has column background_file_id but struct has not related field
2023-05-25T17:44:31.253930061Z: WARNING ▶ [DATABASE] 056 Table projects has column position but struct has not related field
2023-05-25T17:44:31.253949283Z: WARNING ▶ [DATABASE] 057 Table projects has column background_blur_hash but struct has not related field
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xf9186c]

goroutine 1 [running]:
code.vikunja.io/api/pkg/migration.setTeamNamespacesShare(0x132a8a0?, 0x400031f960?)
        /go/src/code.vikunja.io/api/pkg/migration/20221228112131.go:253 +0x14c
code.vikunja.io/api/pkg/migration.init.74.func1(0x14e45a8?)
        /go/src/code.vikunja.io/api/pkg/migration/20221228112131.go:208 +0x2e4
src.techknowlogick.com/xormigrate.(*Xormigrate).runMigration(0x400013cc00, 0x400047b7d0)
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:277 +0xe0
src.techknowlogick.com/xormigrate.(*Xormigrate).migrate(0x400013cc00, {0x0, 0x0})
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:131 +0x338
src.techknowlogick.com/xormigrate.(*Xormigrate).Migrate(...)
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:98
code.vikunja.io/api/pkg/migration.Migrate(0x40000d1500?)
        /go/src/code.vikunja.io/api/pkg/migration/migration.go:68 +0x2c
code.vikunja.io/api/pkg/initialize.FullInit()
        /go/src/code.vikunja.io/api/pkg/initialize/init.go:86 +0xc0
code.vikunja.io/api/pkg/cmd.glob..func24(0x2160900?, {0x14c0010?, 0x0?, 0x0?})
        /go/src/code.vikunja.io/api/pkg/cmd/web.go:71 +0x1c
github.com/spf13/cobra.(*Command).execute(0x2160900, {0x40000a0220, 0x0, 0x0})
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:929 +0x54c
github.com/spf13/cobra.(*Command).ExecuteC(0x2160900)
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x340
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
code.vikunja.io/api/pkg/cmd.Execute()
        /go/src/code.vikunja.io/api/pkg/cmd/cmd.go:44 +0x24
main.main()
        /go/src/code.vikunja.io/api/main.go:22 +0x1c
info: creating the new user vikunja with 1000:1000
usermod: no changes
2023/05/25 17:44:32 No config file found, using default or config from environment variables.
2023-05-25T17:44:32.69085492Z: WARNING  ▶ [DATABASE] 054 Table projects has column identifier but struct has not related field
2023-05-25T17:44:32.690939845Z: WARNING ▶ [DATABASE] 055 Table projects has column background_file_id but struct has not related field
2023-05-25T17:44:32.690964863Z: WARNING ▶ [DATABASE] 056 Table projects has column position but struct has not related field
2023-05-25T17:44:32.690989585Z: WARNING ▶ [DATABASE] 057 Table projects has column background_blur_hash but struct has not related field
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xf9186c]

goroutine 1 [running]:
code.vikunja.io/api/pkg/migration.setTeamNamespacesShare(0x132a8a0?, 0x400065f960?)
        /go/src/code.vikunja.io/api/pkg/migration/20221228112131.go:253 +0x14c
code.vikunja.io/api/pkg/migration.init.74.func1(0x14e45a8?)
        /go/src/code.vikunja.io/api/pkg/migration/20221228112131.go:208 +0x2e4
src.techknowlogick.com/xormigrate.(*Xormigrate).runMigration(0x40004a8360, 0x4000577710)
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:277 +0xe0
src.techknowlogick.com/xormigrate.(*Xormigrate).migrate(0x40004a8360, {0x0, 0x0})
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:131 +0x338
src.techknowlogick.com/xormigrate.(*Xormigrate).Migrate(...)
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:98
code.vikunja.io/api/pkg/migration.Migrate(0x40000d1500?)
        /go/src/code.vikunja.io/api/pkg/migration/migration.go:68 +0x2c
code.vikunja.io/api/pkg/initialize.FullInit()
        /go/src/code.vikunja.io/api/pkg/initialize/init.go:86 +0xc0
code.vikunja.io/api/pkg/cmd.glob..func24(0x2160900?, {0x14c0010?, 0x0?, 0x0?})
        /go/src/code.vikunja.io/api/pkg/cmd/web.go:71 +0x1c
github.com/spf13/cobra.(*Command).execute(0x2160900, {0x40000a0220, 0x0, 0x0})
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:929 +0x54c
github.com/spf13/cobra.(*Command).ExecuteC(0x2160900)
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x340
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
code.vikunja.io/api/pkg/cmd.Execute()
        /go/src/code.vikunja.io/api/pkg/cmd/cmd.go:44 +0x24
main.main()
        /go/src/code.vikunja.io/api/main.go:22 +0x1c
info: creating the new user vikunja with 1000:1000
usermod: no changes
2023/05/25 17:44:34 No config file found, using default or config from environment variables.
2023-05-25T17:44:34.368517612Z: WARNING ▶ [DATABASE] 054 Table projects has column identifier but struct has not related field
2023-05-25T17:44:34.368702777Z: WARNING ▶ [DATABASE] 055 Table projects has column background_file_id but struct has not related field
2023-05-25T17:44:34.368784776Z: WARNING ▶ [DATABASE] 056 Table projects has column position but struct has not related field
2023-05-25T17:44:34.368845905Z: WARNING ▶ [DATABASE] 057 Table projects has column background_blur_hash but struct has not related field
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xf9186c]

goroutine 1 [running]:
code.vikunja.io/api/pkg/migration.setTeamNamespacesShare(0x132a8a0?, 0x4000429960?)
        /go/src/code.vikunja.io/api/pkg/migration/20221228112131.go:253 +0x14c
code.vikunja.io/api/pkg/migration.init.74.func1(0x14e45a8?)
        /go/src/code.vikunja.io/api/pkg/migration/20221228112131.go:208 +0x2e4
src.techknowlogick.com/xormigrate.(*Xormigrate).runMigration(0x4000534d50, 0x40005edc80)
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:277 +0xe0
src.techknowlogick.com/xormigrate.(*Xormigrate).migrate(0x4000534d50, {0x0, 0x0})
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:131 +0x338
src.techknowlogick.com/xormigrate.(*Xormigrate).Migrate(...)
        /go/pkg/mod/src.techknowlogick.com/xormigrate@v1.5.0/xormigrate.go:98
code.vikunja.io/api/pkg/migration.Migrate(0x40001b5400?)
        /go/src/code.vikunja.io/api/pkg/migration/migration.go:68 +0x2c
code.vikunja.io/api/pkg/initialize.FullInit()
        /go/src/code.vikunja.io/api/pkg/initialize/init.go:86 +0xc0
code.vikunja.io/api/pkg/cmd.glob..func24(0x2160900?, {0x14c0010?, 0x0?, 0x0?})
        /go/src/code.vikunja.io/api/pkg/cmd/web.go:71 +0x1c
github.com/spf13/cobra.(*Command).execute(0x2160900, {0x4000192200, 0x0, 0x0})
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:929 +0x54c
github.com/spf13/cobra.(*Command).ExecuteC(0x2160900)
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x340
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
code.vikunja.io/api/pkg/cmd.Execute()
        /go/src/code.vikunja.io/api/pkg/cmd/cmd.go:44 +0x24
main.main()
        /go/src/code.vikunja.io/api/main.go:22 +0x1c
```

It then restarts indefinitely. What could be wrong? Thanks.

### Vikunja Frontend Version

vikunja/frontend unstable 5f53bbc0de97 33 hours ago 63.4MB

### Vikunja API Version

vikunja/api unstable 05b08bf915c3 27 hours ago 58.5MB

### Can you reproduce the bug on the Vikunja demo site?

No

**Comment by Fma965** [comment]
Same here, had to revert to 0.20.4

**Comment by Break-Off** [comment]
@kolaente Here is the command output:

```
MariaDB [VK_database]> select * from team_namespaces where namespace_id not in (select id from namespaces);
+----+---------+--------------+-------+---------------------+---------------------+
| id | team_id | namespace_id | right | created             | updated             |
+----+---------+--------------+-------+---------------------+---------------------+
|  1 |       2 |            8 |     1 | 2022-08-25 16:25:22 | 2022-08-25 16:25:25 |
+----+---------+--------------+-------+---------------------+---------------------+
1 row in set (0.004 sec)
```

Thanks.

**Comment by kolaente** [comment]
Can you delete that entry and try again?

**Comment by Break-Off** [comment]
> Can you delete that entry and try again?

It works now, but the DB is now a mess, at least looking at it from the frontend. All my projects have vanished except nine of them which have duplicated six times, tasks done from a while ago are now undone, and all the tasks don't have a parent project (they only show on the overview).

![image](https://github.com/go-vikunja/api/assets/5843433/4d55d7ff-7c3f-439b-bf94-028d9ecae6e4)

**Comment by Break-Off** [comment]
Hey, I've made a partial dump of the projects table of my DB and it seems that the repeated projects are part of the migration of the offending build:
![image](https://github.com/go-vikunja/api/assets/5843433/f692ba70-40e7-49c7-aa1f-579853ad7da9)

It goes up to 62900 rows (!)

**Comment by kolaente** [comment]
Well, that should not happen.

I tried to reproduce this with a prod dump of my own instance, but it did not duplicate any projects. The migration only created new projects for old namespaces, as it should.

Are the duplicated projects ones which were created for old namespaces, or are they projects which existed before the migration?

I noticed a few little things which caused archived subprojects to show up again, and minor ui issues. Also a bug which caused some subprojects to be unavailable. These should now be fixed. Please update [once the build is done](https://drone.kolaente.de/vikunja/api/5178) again to see if you can now see all projects.

> and all the tasks don't have a parent project (they only show on the overview).

Can you check in the db which project these tasks belong to?

**Comment by variasdesign** [comment]
> Are the duplicated projects ones which were created for old namespaces, or are they projects which existed before the migration?

The first: old namespaces turned to rpojects.

> I noticed a few little things which caused archived subprojects to show up again, and minor ui issues. Also a bug which caused some subprojects to be unavailable. These should now be fixed. Please update [once the build is done](https://drone.kolaente.de/vikunja/api/5178) again to see if you can now see all projects.

Yes, I've got these bugs too, particularly the hidden subprojects. Although I've already moved them out of the hidden parent project and remade it.

> > and all the tasks don't have a parent project (they only show on the overview).
> 
> Can you check in the db which project these tasks belong to?

They belong to the hidden subprojects.

Thanks for your work!

**Comment by kolaente** [comment]
> The first: old namespaces turned to projects.

Where they all created at once or did you run the migration multiple times? Or put the other way, if you restore an old dump from before running the migration and upgrade from there, do you get duplicate projects again?

**Comment by Fma965** [comment]
I'm not sure if you fixed this or i did by cleaning up some random entries in my SQL but it's now working for me on "latest"

**Comment by kolaente** [comment]
@Fma965 latest or unstable? latest is the last release.

**Comment by Fma965** [comment]
Frontend Version: 0.20.5
API Version: v0.20.4

**Comment by kolaente** [comment]
The last release does not have the changes causing the issue.

**Comment by variasdesign** [comment]
> > The first: old namespaces turned to projects.
> 
> Where they all created at once or did you run the migration multiple times? Or put the other way, if you restore an old dump from before running the migration and upgrade from there, do you get duplicate projects again?

Sorry for my lack of response. I've tried to recreate the old dump in unstable but for some reason my MariaDB container didn't want to start and I parked it. I purged the offending namespaces/projects and all is working correctly now. Alas, that doesn't help you much, sorry about that.

**Comment by Fma965** [comment]
> The last release does not have the changes causing the issue.

https://github.com/go-vikunja/api/issues/72#issuecomment-1566140728

I'm confused, i was never on unstable and had this issue for sure on latest when i posted the above comment.

**Comment by kolaente** [comment]
But if you're on the api 0.20.4, that's the last stable release? Which version did you use before that, where the crash happened?

**Comment by kolaente** [comment]
Closing this as it should be fixed, please ping if you still have this problem.
